src/service/position-observer/position-observer-entry.js

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use baseInstance file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  layoutRectEquals,
+  layoutRectsOverlap,
+  layoutRectsRelativePos,
+} from '../../layout-rect';
+import {
+  PositionObserverFidelity,
+  LOW_FIDELITY_FRAME_COUNT,
+} from './position-observer-fidelity';
+import {dev} from '../../log';
+
+/**
+ * TODO (@zhouyx): rename relativePos to relativePositions
+ * The positionObserver returned position value which includes the position rect
+ * relative to viewport. And viewport rect which always has top 0, left 0, and
+ * viewport width and height.
+ * @typedef {{
+ *  positionRect: ?../../layout-rect.LayoutRectDef,
+ *  viewportRect: !../../layout-rect.LayoutRectDef,
+ *  relativePos: string,
+ * }}
+ */
+export let PositionInViewportEntryDef;
+
+export class PositionObserverEntry {
+  /**
+   *
+   * @param {!Element} element
+   * @param {!PositionObserverFidelity} fidelity
+   * @param {!function(?PositionInViewportEntryDef)} handler
+   */
+  constructor(element, fidelity, handler) {
+    /** @const {!Element} */
+    this.element = element;
+
+    /** @const {!function(?PositionInViewportEntryDef)} */
+    this.handler_ = handler;
+
+    /** @type {!PositionObserverFidelity} */
+    this.fidelity = fidelity;
+
+    /** @type {number} */
+    this.turn = (fidelity == PositionObserverFidelity.LOW) ?
+        Math.floor(Math.random() * LOW_FIDELITY_FRAME_COUNT) : 0;
+
+    /** @type {?PositionInViewportEntryDef} */
+    this.prevPosition_ = null;
+  }
+
+  /**
+   * Call to update entry position
+   * @param {!PositionInViewportEntryDef} position
+   */
+  trigger(position) {
+
+    const prevPos = this.prevPosition_ ;
+    if (prevPos
+        && layoutRectEquals(prevPos.positionRect, position.positionRect)
+        && layoutRectEquals(prevPos.viewportRect, position.viewportRect)) {
+      // position didn't change, do nothing.
+      return;
+    }
+
+    dev().assert(position.positionRect);
+    const positionRect =
+        /** @type {!../../layout-rect.LayoutRectDef} */ (position.positionRect);
```

Comment:
This would be cleaner if the `position` we passed to trigger weren't nullable, and the one we pass to handler was. Or, we just pass two objects to the handler.

Reply:
`position` is always NOT nullable. `position.postionRect` is be nullable. 
The problem is I want to pass `position.positionRect = null` to handler for the first time. After that I don't want to invoke handler with another positionRect null.

Reply:
> position.postionRect is be nullable.

That's what I meant.

> The problem is I want to pass position.positionRect = null to handler for the first time.

Don't you mean when there's no intersection? It can still happen that way by using a second type, or passing two arguments to the handler function.

Reply:
emmm. I sorry I kind of get lost here. 
So what you want is have `position.positionRect` NOT nullable when passed to `#trigger()`, but have it nullable when passed to `#handler()`? This is what I have now.

Yes I can definitely change the arguments to the handler function.  But now handler function expects single object, and I think our code is already depending on that return type. Dima proposed some more general object type that we might migrate to. But that's 
following work then.